### ISSUE TITLE
publication.on('subscribed') not fired for new track

### ISSUE THREAD

**Issue by kmahlqvist** [twilio/twilio-video.js#884]
When changing my video track to a new one the publication subscribed event never gets called.

The initial track works fine, and if I try to add the new track a few times it finally works. So the code seems OK, just not getting the subscribed event every time.

It works fine for 2.0.0-beta15 but it does not work for 2.0.0 and 2.1.0

Is there something else I need to do or it is a bug?

**How I do it:**
The initial track is removed, publication.on('unsubscribed') does get called for receiver. Then the new track is added and I get participant.on('trackPublished') on the receiver. But then nothing else happens.

**Code receiver**
```js
participant.on('trackPublished', publication => this.subscribePublication(publication, participant));

private subscribePublication(publication: RemoteTrackPublication, participant: RemoteParticipant) {
            publication.on('subscribed', (track) => { // Not called for new track });
}
```

**Code transmitter**
```js
this.room.localParticipant.unpublishTrack(track);
await this.room.localParticipant.publishTrack(track, { logLevel: 'warn', name: name });
}
```

**Comment by makarandp0** [comment]
@kmahlqvist, Thank you for writing about this issue. Does this issue repro consistently ? Can you please share the room sid (and any client side logs) for the room where the issue was repro'ed.


I also have one question and a speculative suggestion for you to try:
In the `Code transmitter ` does the publishTrack promise resolve ?  Can you try adding some wait between unpublishTrack and publishTrack ?
```
{
    this.room.localParticipant.unpublishTrack(track);
    
    // wait for few seconds before re-publishing track
    await new Promise(resolve => setTimeout(resolve, 3000));
    
    await this.room.localParticipant.publishTrack(track, { logLevel: 'warn', name: name });
    console.log('track published successfully");
}
```

Thanks,
Makarand
JSDK Team

**Comment by kmahlqvist** [comment]
@makarandp0, Thanks for helping out. Here is some more information.

All running 2.1.0

roomSid: "RMfd439fa7a447716619c54cec1a61de3d":
In this room I just started a call, tried to change track. It failed the first time. Tried again then it worked.

roomSid: "RM5eca391cbdcce84f76f398962265444a"
This room added the delay, and verified that the await finished. Tried changing track a bunch of times, worked less then 50%. It always fail when I have a working track and try to add a new. 


Are there any specific client side logs you want to see? I do some simple logging of all the events and this is the output I get there.

*Receiver*
```
-INFO- [Twilio_Room_Establishing_Connection] {name: "4e722a7f9b2d47f78a7c8ce66a2efa5a", tracks: 2}
-INFO- [Twilio_Room_Connected] {roomSid: "RM228ecfdc3f61f06edb455f522b91f960", participants: Array(0)}
-INFO- [Twilio_Room_Participant_Connected] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "addedAudioTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "subscribedAudioTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "addedVideoTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "subscribedVideoTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "startedAudioTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "startedVideoTrack"}

-INFO- [Twilio_Room_Participant_Event] {event: "videoTrackPublished"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "removedVideoTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "unSubscribedVideoTrack"}
-INFO- [Twilio_Room_Participant_Event] {event: "videoTrackUnpublished"}

-INFO- [Twilio_Room_Participant_Event] {event: "videoTrackPublished"}
-INFO- [Twilio_Room_Participant_Event] {event: "videoTrackUnpublished"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "addedVideoTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "subscribedVideoTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "startedVideoTrack"}

-INFO- [Twilio_Room_Participant_Event] {event: "videoTrackPublished"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "removedVideoTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "unSubscribedVideoTrack"}
-INFO- [Twilio_Room_Participant_Event] {event: "videoTrackUnpublished"}

-INFO- [Twilio_Room_Participant_Event] {event: "videoTrackPublished"}
-INFO- [Twilio_Room_Participant_Event] {event: "videoTrackUnpublished"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "addedVideoTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "subscribedVideoTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PAa2730428e92176bb80162d8e1a631ee6", event: "startedVideoTrack"}
```

*Sender*
```
-INFO- [Twilio_Room_Establishing_Connection] {name: "4e722a7f9b2d47f78a7c8ce66a2efa5a", tracks: 2}
-INFO- [Twilio_Room_Connected] {roomSid: "RM228ecfdc3f61f06edb455f522b91f960", participants: Array(1)}
-INFO- [Twilio_Room_Participant_Connected] {participantSid: "PA7f2b2a4050c7d56e7adce2df4cc56f1e"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PA7f2b2a4050c7d56e7adce2df4cc56f1e", event: "addedAudioTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PA7f2b2a4050c7d56e7adce2df4cc56f1e", event: "subscribedAudioTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PA7f2b2a4050c7d56e7adce2df4cc56f1e", event: "addedVideoTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PA7f2b2a4050c7d56e7adce2df4cc56f1e", event: "subscribedVideoTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PA7f2b2a4050c7d56e7adce2df4cc56f1e", event: "startedAudioTrack"}
-INFO- [Twilio_Room_Participant_Event] {participantSid: "PA7f2b2a4050c7d56e7adce2df4cc56f1e", event: "startedVideoTrack"}
-INFO- [Twilio_Room_Local_Event] {event: "startedVideoTrack"}
-INFO- [Twilio_Room_Local_Event] {event: "publishedVideoTrack", trackName: "Screen"}
-INFO- [Twilio_Room_Local_Event] {event: "startedVideoTrack"}
-INFO- [Twilio_Room_Local_Event] {event: "publishedVideoTrack", trackName: "Camera"}
-INFO- [Twilio_Room_Local_Event] {event: "startedVideoTrack"}
-INFO- [Twilio_Room_Local_Event] {event: "publishedVideoTrack", trackName: "Screen"}
-INFO- [Twilio_Room_Local_Event] {event: "startedVideoTrack"}
-INFO- [Twilio_Room_Local_Event] {event: "publishedVideoTrack", trackName: "Camera"}
```

**Comment by makarandp0** [comment]
Hey @kmahlqvist - sorry for the delay on this. I looked into the server logs of rooms you mentioned, but did not find anything obvious. I wasn't able to repro the issue locally either. 

Would it be possible for you to build a sample that demo's the issue ? Perhaps you can modify  [quick-start](https://github.com/twilio/video-quickstart-js/tree/master/quickstart) to make such sample.  

One question - the track you unpublish and publish are different tracks right? Just confirming because in `transmitter` code that you posted -  they appear to be same track
```
this.room.localParticipant.unpublishTrack(track);
await this.room.localParticipant.publishTrack(track, { logLevel: 'warn', name: name });
}
```

One more thing that you could try is to ensure that when you receive `trackPublished` event check if track is already subscribed. if it was already subscribed, you will not receive subscribed event. 

You need to change your receiver code to:
```
participant.on('trackPublished', publication => this.subscribePublication(publication, participant));

private subscribePublication(publication: RemoteTrackPublication, participant: RemoteParticipant) {
 if (!publication.isSubscribed)
      publication.on('subscribed', (track) => { // process track subscription here });
 } else {
      console.log('track is already subscribed to");
     // process track subscription here
 }
```

Thanks,
Makarand

**Comment by jwahdatehagh** [comment]
I think i'm having the same issue as @kmahlqvist.

For me the new publications are not subscribed to when the new track gets published (`publication.isSubscribed === false` and `publication.track === null`) and the `trackSubscribed` / `subscribed` events are never fired.

**Comment by kmahlqvist** [comment]
Sorry for dropping of the radar on this issue. I got a bit swamped. I did however look at it a bit last week and my attempts at recreating the issue in the quick start or a completely new project failed.

So I tried another approach, the current project I'm in is an old angular project that still uses systemjs and the prebuilt dist of twilio-video instead of webpack. And when I quickly moved it over to webpack it somehow worked.

I don't know what about the setups ruins it, but for me it seems like a good excuse to change over to webpack.

So on my part this issue can be closed.

Don't know if this helps you at all @jwahdatehagh, but good luck to you :)
